# Artifact Table 2 for Phase Moducation Side Channel Analysis and JTAG Attacks

This is part of a series of artifacts. Note these notebooks are designed to run on Google Colab so have been simplified from those available in the [full repository](https://github.com/colinoflynn/phase-modulation-sca).

To run these notebooks, simply run from top to bottom.

**WARNING**: Some steps, such as the extraction, can be very slow. Try not to forget you started the process as CoLab will kill unattended notebooks eventually, and you'll have to start again.

In [ ]:
!pip install scared

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 112.9/112.9 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.8/55.8 kB 2.1 MB/s eta 0:00:00


In [ ]:
!wget https://www.dropbox.com/scl/fi/67z8waf8iyf562xjzu9ko/mixers.7z?rlkey=tm97vszf4zodrndwmytynbv5n&dl=1

--2024-07-26 18:29:26--  https://www.dropbox.com/scl/fi/67z8waf8iyf562xjzu9ko/mixers.7z?rlkey=tm97vszf4zodrndwmytynbv5n
Resolving www.dropbox.com (www.dropbox.com)... 162.125.5.18, 2620:100:601d:18::a27d:512
Connecting to www.dropbox.com (www.dropbox.com)|162.125.5.18|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://uc638029a66e5b29b8bf726df425.dl.dropboxusercontent.com/cd/0/inline/CXfWwW_RWz7-qGwPgHCBz1vqYy6sN5r8F6YlWLdtqsufLxPGAalQFA4l5DisnwmI-Q6ARfdtMb28qzUeSTqdA4UUlEs1KbOgCN46npV17AikEEwwdXBRZRJgcaZJ9IT3yFKg2sEdyaLPNoILMtnKom28/file# [following]
--2024-07-26 18:29:27--  https://uc638029a66e5b29b8bf726df425.dl.dropboxusercontent.com/cd/0/inline/CXfWwW_RWz7-qGwPgHCBz1vqYy6sN5r8F6YlWLdtqsufLxPGAalQFA4l5DisnwmI-Q6ARfdtMb28qzUeSTqdA4UUlEs1KbOgCN46npV17AikEEwwdXBRZRJgcaZJ9IT3yFKg2sEdyaLPNoILMtnKom28/file
Resolving uc638029a66e5b29b8bf726df425.dl.dropboxusercontent.com (uc638029a66e5b29b8bf726df425.dl.dropboxusercontent.com)... 162.125.5.15, 2620:100:

In [ ]:
!7z x mixers.7z?rlkey=tm97vszf4zodrndwmytynbv5n


7-Zip [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_US.UTF-8,Utf16=on,HugeFiles=on,64 bits,2 CPUs Intel(R) Xeon(R) CPU @ 2.20GHz (406F0),ASM,AES-NI)

Scanning the drive for archives:
  0M Scan         1 file, 1542379774 bytes (1471 MiB)

Extracting archive: mixers.7z?rlkey=tm97vszf4zodrndwmytynbv5n
--
Path = mixers.7z?rlkey=tm97vszf4zodrndwmytynbv5n
Type = 7z
Physical Size = 1542379774
Headers Size = 607
Method = LZMA2:22
Solid = +
Blocks = 11

  0%      0% 1 - mixers/mixer_bothclks_defaultfi . 20msps_ttest_10k_tvla_group1.npy                                                                             1% 1 - mixers/mixer_bothclks_defaultfi . 20msps_ttest_10k_tvla_group1.npy                   

In [ ]:
!rm -rf  mixers.7z?rlkey=tm97vszf4zodrndwmytynbv5n

In [ ]:
import scared
import estraces
import numpy as np

In [ ]:
knownkey = [43, 126, 21, 22, 40, 174, 210, 166, 171, 247, 21, 136, 9, 207, 79, 60]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
def find_success(att, knownkey):
    recovered_masterkey = np.argmax(att.scores, axis=0).astype('uint8')
    print(recovered_masterkey == knownkey)

    for i in range(0, np.shape(att.convergence_traces)[2]):
        if np.all(np.argmax(att.convergence_traces[:,:,i], axis=0) == knownkey):
            return (i+1)*att.convergence_step
    return -1

In [ ]:
import scared
from scipy import signal


def do_cpa(filename, step=25, window=None, hpf=None, lpf=None):
    ths = estraces.read_ths_from_ets_file(filename)
    print(ths)

    selection_function = scared.aes.selection_functions.encrypt.FirstSubBytes()

    att = scared.CPAAttack(selection_function=selection_function,
                           model=scared.HammingWeight(),
                           discriminant=scared.maxabs,
                           convergence_step=step)

    processes = []

    # Simple high pass filter (HPF)
    if hpf:
        sos = signal.butter(5, hpf, 'highpass', output='sos')

        @scared.preprocess
        def hpfproc(traces):
            return signal.sosfilt(sos, traces)

        processes.append(hpfproc)

    # Simple low pass filter (LPF)
    if lpf:
        sos = signal.butter(5, lpf, 'lowpass', output='sos')

        @scared.preprocess
        def lpfproc(traces):
            return signal.sosfilt(sos, traces)

        processes.append(lpfproc)

    # This resync preprocessing isn't really used, was used with some JTAG captures only
    @scared.preprocess
    def resync(traces):
        import numpy as np
        first = []
        for w in traces:
            first.append(w[0])
        avg = np.mean(first)

        tracesresync = []
        for w in traces:
            if w[0] < avg:
                tracesresync.append(w[:-1])
            else:
                tracesresync.append(w[1:])
        return np.array(tracesresync)
    #processes.append(resync)

    if window:
        container = scared.Container(ths, frame=range(window[0], window[1]), preprocesses=processes)
    else:
        container = scared.Container(ths, preprocesses=processes)


    att.run(container)

    import numpy as np

    recovered_masterkey = np.argmax(att.scores, axis=0).astype('uint8')

    return att

In [ ]:
# Datasets in baseline_traces_cpa online
filenames = [('mixers/mixer_default_120mhz_sync_120msps_cpa.ets', 25, None, None, None),
             ('mixers/mixer_slowerfilter_120mhz_sync_120msps_cpa.ets', 25, None, None, None),
             ('mixers/opticalisolator_mixer_120mhz_sync_120msps_cpa.ets', 25, None, None, None),
             ('mixers/vdiv_120mhz_sync_120msps_cpa.ets', 25, None, 0.1, None),
            ]

## Doing all the CPA attacks on mixer measurements

The following will do all the attacks on *mixer* measurements. The results get saved to a `.npy` file for further use. The `.npy` files have also been saved from prior runs.

In [ ]:
for f in filenames:
    att = do_cpa(f[0], f[1], f[2], f[3], f[4])
    successnum = find_success(att, knownkey)
    print("{:s} = {:d}".format(f[0], successnum))
    np.save(f[0] + ".npy", att)

Trace Header Set:
Name.............: ETS Format THS
Reader...........: ETS format reader of file mixers/mixer_default_120mhz_sync_120msps_cpa.ets with 5000 traces.
ciphertext.......: uint8
key..............: uint8
plaintext........: uint8

[ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True]
mixers/mixer_default_120mhz_sync_120msps_cpa.ets = 3900
Trace Header Set:
Name.............: ETS Format THS
Reader...........: ETS format reader of file mixers/mixer_slowerfilter_120mhz_sync_120msps_cpa.ets with 5000 traces.
ciphertext.......: uint8
key..............: uint8
plaintext........: uint8

[ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True]
mixers/mixer_slowerfilter_120mhz_sync_120msps_cpa.ets = 3400
Trace Header Set:
Name.............: ETS Format THS
Reader...........: ETS format reader of file mixers/vdiv_120mhz_sync_120msps_cpa.ets with 40000 traces.
ciphertext.......: uint8
key..............: 

The above results in a bunch of output, but the important parts are the lines that look like this:

```
mixers/mixer_default_120mhz_sync_120msps_cpa.ets = 3900
mixers/mixer_slowerfilter_120mhz_sync_120msps_cpa.ets = 3400
mixers/opticalisolator_mixer_120mhz_sync_120msps_cpa.ets = 2775
mixers/vdiv_120mhz_sync_120msps_cpa.ets = 25775
```

The above can be found in Table 2, showing the CPA results:

| Measurement                      | Isolator | PLL Filter | TVLA   | CPA   |
|----------------------------------|----------|------------|--------|-------|
| Mixer                            | IC-Based | Default    | 21.7   | 3900  |
| Mixer                            | IC-Based | Lowered    | 20.7   | 3400  |
| Mixer                            | Optical  | Lowered    | 24.1   | 2775  |
| Voltage Divider                  | IC-Based | Lowered    | 4.7    | 25775 |

We'll recreate the T-Test results below.

You may wish to save the .npy files. While we have a copy of them, for a "complete" artifact recreation you should copy the .npy files to the CPA plotting notebook.

The easiest way to do this if using colab (and not a local Jupyter install) is to copy them to a directory & compress them. Then download the file in the CoLab web interface.

The following block will move them to a `results` directory, then the block after that will compress that directory. Look for a `results.7z` file you download.

In [ ]:
%%bash
mkdir results
mv baseline_traces_cpa/*.npy results/.

In [ ]:
!7z a results.7z results

## T-Test Results

The following recreats the T-Test results for Table 2.

In [ ]:
import os
import numpy as np
def plot_t(t_val, N, titledata=""):
    import matplotlib.pylab as plt
    ldata = len(t_val)
    plt.plot(t_val)
    plt.plot([0, ldata], [4.5, 4.5], 'k--')
    plt.plot([0, ldata], [-4.5, -4.5], 'k--')
    plt.title("TVLA Results, N=%d, %s"%(N, titledata))
    plt.xlabel("Sample Number")
    plt.ylabel("T-Test Value")

#Set root data dir, not "mixers"
data_dir = "."

def npload(filename):
    return np.load(os.path.join(data_dir, filename), allow_pickle=True)

def nptload(prefix):
    group1 = npload(prefix + "_tvla_group1.npy")
    group2 = npload(prefix + "_tvla_group2.npy")
    N = round((len(group1) + len(group2)) / 2)
    return group1, group2, N

def npsave(filename, array):
    np.save(os.path.join(data_dir, filename), array)

def nptsave(prefix, group1, group2):
    npsave(prefix + "_tvla_group1.npy", group1)
    npsave(prefix + "_tvla_group2.npy", group2)


In [ ]:
files = [
         [r"mixers/mixer_bothclks_defaultfilter_120mhz_sync_120msps_ttest_10k", 100, 10000, None, None],
         [r"mixers/mixer_bothclks_slowerfilter_120mhz_sync_120msps_ttest_10k", 100, 10000, None, None,],
         [r"mixers/vdiv_120mhz_sync_120msps_ttest_10k", 100, 10000, 0.1, 0.5],
         [r"mixers/opticalisolator_mixer_bothclks_120mhz_sync_120msps_ttest_10k", 100, 10000, None, None,],
        ]

In [ ]:
from scipy.stats import ttest_ind
import matplotlib.pylab as plt
from scipy import signal

def find_max_t(file, start=0, end=-1, hpf=None, lpf=None, resync_jtag=False):
    group1, group2, N = nptload(file)

    if resync_jtag:
        first = []
        for g in group1:
            first.append(g[0])
        avg = np.mean(first)
        group1resync = []
        for g in group1:
            if g[0] > avg:
                group1resync.append(g[:-1])
            else:
                group1resync.append(g[1:])

        group2resync = []
        for g in group2:
            if g[0] > avg:
                group2resync.append(g[:-1])
            else:
                group2resync.append(g[1:])

        group1 = group1resync
        group2 = group2resync

    if hpf:
        sos = signal.butter(5, hpf, 'highpass', output='sos')
        group1 = signal.sosfilt(sos, group1)
        group2 = signal.sosfilt(sos, group2)

    if lpf:
        sos = signal.butter(5, lpf, 'lowpass', output='sos')
        group1 = signal.sosfilt(sos, group1)
        group2 = signal.sosfilt(sos, group2)

    t_val = ttest_ind(group1[start:end], group2[start:end], axis=0, equal_var=False)[0]
    max_t = max(abs(t_val))
    return max_t, t_val
for f in files:
    max_t, t_val = find_max_t(f[0], start=f[1], end=f[2], hpf=f[3], lpf=f[4])
    print("{:s}: {:f}".format(f[0], max_t))

mixers/mixer_bothclks_defaultfilter_120mhz_sync_120msps_ttest_10k: 21.650953
mixers/mixer_bothclks_slowerfilter_120mhz_sync_120msps_ttest_10k: 20.685422
mixers/vdiv_120mhz_sync_120msps_ttest_10k: 4.661121
mixers/opticalisolator_mixer_bothclks_120mhz_sync_120msps_ttest_10k: 24.081051


This outputs the T-Test results. You can confirm the values match the table above (Table 2 in paper):

* mixers/mixer_bothclks_defaultfilter_120mhz_sync_120msps_ttest_10k: 21.650953
* mixers/mixer_bothclks_slowerfilter_120mhz_sync_120msps_ttest_10k: 20.685422
* mixers/vdiv_120mhz_sync_120msps_ttest_10k: 4.661121
* mixers/opticalisolator_mixer_bothclks_120mhz_sync_120msps_ttest_10k: 24.081051

This should now be Table 2 fully reproduced, both CPA and T-Test values.